pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/pkg/integrations/v2"
+	"github.com/grafana/agent/pkg/logs"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	cacheFileMode = 0600
+)
+
+// EventHandler watches for Kubernetes Event objects and hands them off to
+// Agent's logs subsystem (embedded promtail).
+type EventHandler struct {
+	LokiClient    *logs.Instance
+	Log           log.Logger
+	CacheFile     *os.File
+	CachePath     string
+	LastEvent     *ShippedEvents
+	InitEvent     *ShippedEvents
+	EventInformer cache.SharedIndexInformer
+	ClusterLabel  string
+	SendTimeout   time.Duration
+	sync.Mutex
+}
+
+// ShippedEvents stores a timestamp and map of event ResourceVersions shipped for that timestamp.
+// Used to avoid double-shipping events upon restart.
+type ShippedEvents struct {
+	// shipped event's timestamp
+	Timestamp time.Time `json:"ts"`
+	// map of event RVs (resource versions) already "shipped" (handed off) for this timestamp.
+	// this is to handle the case of a timestamp having multiple events,
+	// which happens quite frequently.
+	RvMap map[string]struct{} `json:"resourceVersion"`
+}
+
+func newEventHandler(l log.Logger, globals integrations.Globals, c *Config) (integrations.Integration, error) {
+	var (
+		config *rest.Config
+		err    error
+	)
+
+	if c.InCluster {
+		config, err = rest.InClusterConfig()
+	} else {
+		config, err = clientcmd.BuildConfigFromFlags("", c.KubeconfigPath)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// get an informer
+	factory := informers.NewSharedInformerFactory(clientset, time.Duration(c.InformerResync)*time.Second)
+	eventInformer := factory.Core().V1().Events().Informer()
+
+	eh := &EventHandler{
+		LokiClient:    globals.Logs.Instance(c.LogsInstance),
+		Log:           l,
+		CachePath:     c.CachePath,
+		EventInformer: eventInformer,
+		ClusterLabel:  c.ClusterName,
+		SendTimeout:   time.Duration(c.SendTimeout) * time.Second,
+	}
+	// set the resource handler fns
+	eh.initInformer(eventInformer)
+	return eh, nil
+}
+
+// Initialize informer by setting event handler fns
+func (eh *EventHandler) initInformer(eventsInformer cache.SharedIndexInformer) {
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.addEvent,
+		UpdateFunc: eh.updateEvent,
+		DeleteFunc: eh.deleteEvent,
+	})
+}
+
+// Handles new event objects
+func (eh *EventHandler) addEvent(obj interface{}) {
+	e, ok := obj.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", obj)
+		return
+	}
+	eventTs := getTimestamp(e)
+
+	// if event is older than the one stored in cache on startup, we've shipped it
+	if eventTs.Before(eh.InitEvent.Timestamp) {
```

Comment:
Is there a race condition here? Is it possible for this to be called before `eh.InitEvent` is ever set?

Reply:
I don't think so. `eh.InitEvent` is set [here](https://github.com/grafana/agent/pull/1310/files#diff-cb8f2bca76034c1cfc71cd9e77f19f06274accfc7cc89e87ffef10c0bb65cceaR369) and the informer gets started [here](https://github.com/grafana/agent/pull/1310/files#diff-cb8f2bca76034c1cfc71cd9e77f19f06274accfc7cc89e87ffef10c0bb65cceaR413). A `ResourceEventHandlerFunc` will only get called once the informer is started, unless there's some sort of closure behavior I am misunderstanding...